include/clap/ext/draft/undo-redo.h

```diff
@@ -0,0 +1,87 @@
+/// whole plugin state but rather only the necessary data for undoing/redoing a single action
+/// on top of the current plugin state.
+
+static const char CLAP_EXT_UNDO_REDO[] = "clap.undo-redo.draft/0";
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// Information about a single incremental change event.
+// The host can use context and action to embed a message into its own undo history.
+typedef struct clap_incremental_change_event_description {
+   // A brief, human-readable description of the event context, for example:
+   //   - "Filter 1 > Cutoff"
+   //   - "Param Seq A"
+   // It should not contain the name of the plugin.
+   char context[CLAP_NAME_SIZE];
+   // A brief, human-readable description of what happened, for example:
+   //   - "Set to 100%"
+   //   - "Steps edited"
+   char action[CLAP_NAME_SIZE];
+} clap_incremental_change_event_description_t;
+
+typedef struct clap_plugin_undo_redo{
+   // Pulls an incremental change event object into stream.
+   // The plugin must provide an event description together with the object.
+   // If the host still holds a redo event object, it should compare the received
+   // undo event object with the next redo object. If it is not equal, the host may
+   // omit the next and all following redo event objects for that plugin.
+   // before (besides it wants to support undo history branching).
+   // Returns true if the incremental change event object was correctly saved.
+   // [main-thread]
+   bool(CLAP_ABI *pull_undo_event_object)(const clap_plugin_t *plugin,
+                                          const clap_ostream_t *stream,
+                                          clap_incremental_change_event_description_t* description);
+
+   // Applies an incremental change event object from stream in order to undo the corresponding
+   // plugin action.
+   // The host should move the incremental change event object onto its redo stack.
+   // Returns true if the incremental change event object was correctly applied.
+   // The plugin must not call notify_undo as this is implied by returning true already.
+   // [main-thread]
+   bool(CLAP_ABI *perform_undo)(const clap_plugin_t *plugin,
+                                const clap_istream_t *stream);
+
+   // Applies an incremental change event object from stream in order to redo the corresponding
+   // plugin action.
+   // The host should move the incremental change event object onto its undo stack.
+   // Returns true if the incremental change event object was correctly applied.
+   // The plugin must not call mark_dirty_for_undo as this is implied by returning true already.
+   // [main-thread]
+   bool(CLAP_ABI *perform_redo)(const clap_plugin_t *plugin,
+                                const clap_istream_t *stream);
+} clap_plugin_undo_redo_t;
+
+typedef struct clap_host_undo_redo {
+   // Tell the host that the plugin state has changed internally and

```

Comment:
I hope I could resolve some of those concerns with the 'pending objects' concept. This means, the plugin has to buffer pending objects and the host must sync all pending objects first before it can perform an undo/redo on the plugin.

I think the buffering shouldn't be much of a problem for the plugin as it already has those objects for its internal undo/redo. So it just needs to store them a bit longer until the host has pulled them.